### ISSUE TITLE
Incorrect F821 raised for list comprehension in `Annotated`

### ISSUE THREAD

**Issue by Harry-Lees** [astral-sh/ruff#7879]
Using Ruff `0.0.292` on the following code snippet

```python
from typing import Annotated

foo = [1, 2, 3, 4, 5]

class Bar:
    baz: Annotated[
        str,
        [qux for qux in foo],
    ]
```

The following error is raised:

```
test.py:8:10: F821 Undefined name `qux`
Found 1 error.
```

The minimal reproducable example looks a little bit weird, but I came across this when writing a [pydantic validator](https://docs.pydantic.dev/latest/concepts/validators/#annotated-validators) on some (IMO) reasonable pydantic code. A more sensible example can be seen below.

```python
from typing import Annotated

from pydantic import BaseModel
from pydantic.functional_validators import BeforeValidator

foo = [1, 2, 3, 4, 5]

class Bar(BaseModel):
    baz: Annotated[
        list[int], BeforeValidator(lambda values: [value**2 for value in values])
    ]

qux = Bar(baz=foo)
```

```
test.py:10:52: F821 Undefined name `value`
Found 1 error.
```

**Comment by charliermarsh** [comment]
Thanks!

**Comment by charliermarsh** [comment]
The cause here isn't immediately obvious to me, but definitely looks like a bug.

**Comment by charliermarsh** [comment]
I see the error here.
